### DISCUSSION TITLE
Any  Silicon M1 native (no Rosetta 2) support?

### DISCUSSION THREAD

**Discussion by josephschito** [jamulussoftware/jamulus#2148] [Ideas]

**Comment by ann0see** [comment]
Maybe related to #1795 and https://github.com/jamulussoftware/jamulus/pull/2299
QT6 support is being worked on by multiple contributors.

**Comment by dcorson-ticino-com** [comment]
> Any  Silicon M1 native (no Rosetta 2) support?

If you know how, all help is welcome !

**Comment by hoffie** [comment]
Currently, there's no such support.

Github Actions, where we build our binaries, does not provide arm64/M1 runners at the moment (and I've seen no plans to do so publicly soon, probably mainly due to lack on Azure support).

I've done some quick tests wrt cross-compiling on a private branch.
`QMAKE_APPLE_DEVICE_ARCHS="arm64" QT_ARCH="arm64"` allows basic building, but [fails linking](https://github.com/hoffie/jamulus/runs/5077839462?check_suite_focus=true) as we lack Qt arm64 binaries and I don't see them on Qt mirrors either. I'm not sure, but we might have to get them elsewhere or build them from source. The latter should be possible, but takes (CI) time as well.

It also seems like it would be best (maybe even required?) to build on macos-11 and with some recent Xcode. Both is being worked on (PRs are open), but there is lots of interconnected stuff which needs to be tackled first.

Edit:
> You can try out Qt on Apple Silicon already now by installing the 6.2 preview from the Qt installer. The Qt SDK is fully universal, and should run on both Intel and Apple Silicon hardware.

[Looks like 6.2 should have it.](https://www.qt.io/blog/qt-on-apple-silicon)
Update: Qt 6.2 works, see below.

**Comment by hoffie** [comment]
`Jamulus: Mach-O 64-bit arm64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|PIE>`

So... build succeeded and the output is arm64. Does anyone want to try if it works? How can it be confirmed that this does not run via Rosetta? Is the binary arch indication by `file` sufficient to prove this?

[**> Download**](https://github.com/hoffie/jamulus/suites/5189988832/artifacts/157995502)

The relevant tree is here: https://github.com/hoffie/jamulus/tree/autobuild-poc-mac-arm64e
It's basically:
- Qt6 support (#2299)
- macos-11 support via switch to create-dmg (#2207)
- and a single commit to set the proper build args: https://github.com/hoffie/jamulus/commit/79945e6887cd7455014f6885311ecbbb2b536e14

This test build was perfomed with Xcode 13. I guess, earlier versions might work as well (though some 12.x, I guess).

If this works as well (please do give feedback! - @josephschito? :)) it might be realistic to include this for 3.9.0 once the other PRs have landed.
The next step would probably be a unified build, but separate builds would probably be sufficient for now.

**Reply by josephschito** [reply]
I tried it but...
<img width="262" alt="Screenshot 2022-02-07 at 09 42 57" src="https://user-images.githubusercontent.com/4095923/152753860-54d0c5b8-d068-46e7-bc6e-5f650361150c.png">

**Reply by hoffie** [reply]
@josephschito 
Thanks for the feedback. After some research, we might hit one of the following problems:
- Seems like [macOS 11 on M1 requires all binaries to be signed](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-universal-apps-release-notes/). We only sign production builds, therefore, my test build was not signed. Seems like one can work around this like that: `sudo xattr -r -d com.apple.quarantine Jamulus.app` - could you try that?
- There are [hints that the .dmg has to be in APFS format (HFS+ will not work)](https://github.com/electron-userland/electron-builder/issues/5850#issuecomment-841760516). I'm currently checking what kind of .dmg was created.

Just to confirm, at what stage do you get this error message? Is it when opening the .dmg or is the .dmg opening successful and the issue occurs when attempting to start the actual binary?

**Reply by josephschito** [reply]
It works! (I didn't yet try all the functionalities)
<img width="657" alt="Screenshot 2022-02-07 at 11 33 35" src="https://user-images.githubusercontent.com/4095923/152771695-7568dc9a-c6ee-4283-975a-e865f295b072.png">
<img width="1444" alt="Screenshot 2022-02-07 at 11 33 31" src="https://user-images.githubusercontent.com/4095923/152771703-33eee425-78e0-4403-8285-989607092e2c.png">

**Reply by hoffie** [reply]
Glad to hear! Please do report how further testing/usage turns out! :)

**Reply by hoffie** [reply]
I've opened #2357 for proper tracking of getting this integrated.

**Reply by ashstrahle** [reply]
So excited for this. Well done. Performance is great. 
Minor bug so far. Profile Country/Region index seems to be out by one. i.e Needed to select Aruba, to get Australian flag in mixer. 

Looking forward to this getting released.

<img width="539" alt="image" src="https://user-images.githubusercontent.com/54830016/156573029-6b1c260b-215f-45ac-88a6-0e6d4b971308.png">

**Reply by pljones** [reply]
> jamulus_3.8.2beta1dev-9c1db57_mac.dmg 

You'll want that renamed unless it's a very fat dmg :).

**Reply by hoffie** [reply]
> So excited for this. Well done. Performance is great.

Thanks for the feedback, that's great to hear! Are you seeing some explicit performance improvements, i.e. better ping times, less audio drop outs or something?

> Minor bug so far. Profile Country/Region index seems to be out by one. i.e Needed to select Aruba, to get Australian flag in mixer.

There's a [known deficit in the Qt6 compatibility code around the country code handling](https://github.com/jamulussoftware/jamulus/pull/2299#issuecomment-1034109767). I might ping you for re-testing once we think it's sorted out. :)

**Reply by ashstrahle** [reply]
Interface response is much smoother and quicker. My ping times are equally woeful due to my location. Can set Audio Quality to High, with 5.33ms Buffer Delay, and Enable Small Network Buffers without any drop outs.

Certainly, reach out if you need further testing.

**Comment by chrisrimple** [comment]
I haven't had an opportunity to test yet, but will try during the coming days.  Was there an answer to the question "how do we ensure that Rosetta2 isn't being used?"

**Reply by ashstrahle** [reply]
Right there. Kind: Apple (not Intel)

<img width="753" alt="image" src="https://user-images.githubusercontent.com/54830016/156902704-47703d2e-4993-4ee8-a201-9703538984b7.png">

**Reply by ashstrahle** [reply]
The native M1 compile has been working exceptionally well. Would be great to see it released with 3.9.0. PR coming soon?

**Reply by dcorson-ticino-com** [reply]
I believe it runs great, the ARM64 hardware really flys.  
I've been running an ARM64 under Linux for the last year.  
The performance is impressing.

**Reply by ann0see** [reply]
Yes, there's a lot going on concerning the autobuild logic and if you're lucky the ARM build is finished in the next release. @hoffie opened an issue: https://github.com/jamulussoftware/jamulus/issues/2357 feel free to participate there.

**Reply by hoffie** [reply]
> PR coming soon?

The PR itself is trivial, the ground work before wasn't/isn't. ;)
#2299 is what needs to be done next, I'm hoping to continue to work on it in the coming days.

There's still a chance to make it into 3.9.0.

**Comment by PeRAux** [comment]
the download link doesn't work anymore. Any way to get the build for testing?

**Reply by hoffie** [reply]
I've rebased the Qt6 PR (#2299) to latest master, have pushed a potential fix for the country code migration problem between Jamulus/Qt5 (= current Mac builds) and Jamulus/Qt6 (= M1 build) and have re-done the required build changes in a potentially mergable way:
[:arrow_down: jamulus_3.8.2dev-51fea25_mac_arm64.dmg](https://github.com/hoffie/jamulus/suites/5946184264/artifacts/203841762)

@ashstrahle Can you check if the country selection bug is fixed for you?
Note: If you update from the earlier M1 build linked above, this new build is likely going to break the country code interpretation once more. A useful check would be: Start an official Jamulus Mac build (= Qt5), select the proper Country & exit cleanly. Then, start the new M1 build linked from this comment and verify that it continues showing the proper country.

**Reply by ashstrahle** [reply]
Have tested this morning. Country code selection is working perfectly. Have tried changing different countries and all working as expected. Nice work. Go for launch ;)

**Reply by PeRAux** [reply]
Thanks! Seems to run smoothly! Will be properly tested friday evening.
Great work!

**Comment by ann0see** [comment]
We have a Pull Request open. Could you please test and comment if you find issues here: https://github.com/jamulussoftware/jamulus/pull/2682?

**Reply by PeRAux** [reply]
I couldn't find any problems running the native M1 version. Both server and client are running just fine on my Mac!

**Reply by josephschito** [reply]
Idem! LGTM! 👏 🙌
